coqlspclient/proof_state.py

```diff
@@ -262,9 +263,45 @@ def __search_notation(call):
 
         return traverse_ast(self.__current_step.ast.span)
 
+    def __get_last_term(self):
+        terms = self.coq_file.terms
+        if len(terms) == 0:
+            return None
+        last_term = terms[0]
+        for term in terms[1:]:
+            if (term.ast.range.end.line > last_term.ast.range.end.line) or (
+                term.ast.range.end.line == last_term.ast.range.end.line
+                and term.ast.range.end.character > last_term.ast.range.end.character
+            ):
+                last_term = term
+        return last_term
+
+    def __get_program_context(self):
+        tag = CoqFile.expr(self.__current_step.ast)[1][1]
+        if tag == 1:
+            # Obligation N of id
+            id = CoqFile.expr(self.__current_step.ast)[2][1][2][0][1][1]
+            return self.__program_context[id]
+        elif tag in [3, 5]:
+            # Next Obligation / Obligation N
+            id = self.coq_file.current_goals().program[0][0][1]
+            return self.__program_context[id]
+        else:
+            return None
```

Comment:
Since we are not sure about the tags, I think we should throw a RuntimeError here. In this way, in the future we can find a missing tag more easily